include/openPMD/regions/Box.hpp

```diff
@@ -0,0 +1,513 @@
+      assert(r <= b1 && !(r <= b2));
+      vol += r.size();
+    }
+    using std::max;
+    assert(vol >= max(size_type(0), b1.size() - b2.size()) && vol <= b1.size());
+    for (size_t i = 0; i < rs.size(); ++i)
+      for (size_t j = i + 1; j < rs.size(); ++j)
+        assert(isdisjoint(rs[i], rs[j]));
+#endif
+    return rs;
+  }
+  friend std::vector<Box> difference(const Box &b1, const Box &b2) {
+    return b1 - b2;
+  }
+
+  friend std::vector<Box> operator|(const Box &b1, const Box &b2) {
+    auto rs = b1 - b2;
+    if (!b2.empty())
+      rs.push_back(b2);
+#if REGIONS_DEBUG
+    // Postcondition
+    size_type vol = 0;
+    for (const auto &r : rs) {
+      assert(!r.empty());
+      assert(r <= b1 || r <= b2);
+      vol += r.size();
+    }
+    assert(vol >= b1.size() && vol <= b1.size() + b2.size());
+    for (size_t i = 0; i < rs.size(); ++i)
+      for (size_t j = i + 1; j < rs.size(); ++j)
+        assert(isdisjoint(rs[i], rs[j]));
+#endif
+    return rs;
+  }
+  friend std::vector<Box> setunion(const Box &b1, const Box &b2) {
+    return b1 | b2;
+  }
+
+  friend std::vector<Box> operator^(const Box &b1, const Box &b2) {
+    auto rs = b1 - b2;
+    auto rs1 = b2 - b1;
+    // TODO: Avoid this concatenation
+    rs.insert(rs.end(), rs1.begin(), rs1.end());
+#if REGIONS_DEBUG
+    // Postcondition
+    size_type vol = 0;
+    for (const auto &r : rs) {
+      assert(!r.empty());
+      assert((r <= b1) ^ (r <= b2));
+      vol += r.size();
+    }
+    using std::abs;
+    assert(vol >= abs(b1.size() - b2.size()) && vol <= b1.size() + b2.size());
+    for (size_t i = 0; i < rs.size(); ++i)
+      for (size_t j = i + 1; j < rs.size(); ++j)
+        assert(isdisjoint(rs[i], rs[j]));
+#endif
+    return rs;
+  }
+  friend std::vector<Box> symmetric_difference(const Box &b1, const Box &b2) {
+    return b1 ^ b2;
+  }
+
+  /** Output a box
+   */
+  friend std::ostream &operator<<(std::ostream &os, const Box &b) {
+    return os << "(" << b.lo << ":" << b.hi << ")";
+  }
+};
+
+} // namespace Regions
+} // namespace openPMD
+
+namespace std {
+
+template <typename T, std::size_t D>
+struct equal_to<openPMD::Regions::Box<T, D>>;
+
+template <typename T> struct equal_to<openPMD::Regions::Box<T, 0>> {
+  constexpr bool operator()(const openPMD::Regions::Box<T, 0> &x,
+                            const openPMD::Regions::Box<T, 0> &y) const {
+    return x.empty() == y.empty();
+  }
+};
+
+template <typename T, std::size_t D>
+struct equal_to<openPMD::Regions::Box<T, D>> {
+  constexpr bool operator()(const openPMD::Regions::Box<T, D> &x,
+                            const openPMD::Regions::Box<T, D> &y) const {
+    if (x.empty() && y.empty())
+      return true;
+    if (x.empty() || y.empty())
+      return false;
+    return openPMD::Regions::helpers::tuple_eq(
+        make_tuple(x.lower(), x.upper()), make_tuple(y.lower(), y.upper()));
+  }
+};
+
+template <typename T, std::size_t D> struct hash<openPMD::Regions::Box<T, D>> {
```

Comment:
Just curious: what are these hashes needed for in terms of functionality?

Reply:
Providing a hash function allows using these types as keys in some data structures, e.g. `unordered_map` or `unordered_set`. Hash functions need to be compatible with the equality operator. Other data structures (`map`, `set`) use comparison operators (`std::less`) for their keys.